subsys/net/l2/canbus/canbus.c

```diff
@@ -1002,6 +1082,191 @@ static enum net_verdict process_frame(struct net_pkt *pkt)
+		NET_DBG("Frame is for Ethernet only %02x:%02x:%02x:%02x:%02x:%02x",
+			lladdr->addr[0], lladdr->addr[1], lladdr->addr[2],
+			lladdr->addr[3], lladdr->addr[4], lladdr->addr[5]);
+		return NET_CONTINUE;
+	}
+
+	net_pkt_cursor_init(pkt);
+	/*Forward all broadcasts*/
+	if (net_eth_is_addr_broadcast((struct net_eth_addr *)lladdr->addr) ||
+	    net_eth_is_addr_multicast((struct net_eth_addr *)lladdr->addr)) {
+		clone_pkt = net_pkt_shallow_clone(pkt, NET_CAN_ALLOC_TIMEOUT);
+		if (clone_pkt) {
+			eth_to_can_lladdr(clone_pkt, true);
+			NET_DBG("Frame is %scast %02x:%02x:%02x:%02x:%02x:%02x,"
+				" CAN dest: 0x%x",
+				net_eth_is_addr_broadcast(
+					(struct net_eth_addr *)lladdr->addr) ? "broad" :
+										"multi",
+				lladdr->addr[0], lladdr->addr[1], lladdr->addr[2],
+				lladdr->addr[3], lladdr->addr[4], lladdr->addr[5],
+				get_dest_lladdr(clone_pkt));
+			net_pkt_set_family(clone_pkt, AF_INET6);
+			forward_eth_frame(clone_pkt);
+		} else {
+			NET_ERR("PKT forwarding: cloning failed");
+		}
+
+		return NET_CONTINUE;
+	}
+
+	/*This frame is for 6LoCAN only*/
+	eth_to_can_lladdr(pkt, false);
+	net_pkt_set_family(pkt, AF_INET6);
+	forward_eth_frame(pkt);
+	NET_DBG("Frame is for CANBUS %02x:%02x:%02x:%02x:%02x:%02x, "
+		"CAN dest: 0x%x",
+		lladdr->addr[0], lladdr->addr[1], lladdr->addr[2],
+		lladdr->addr[3], lladdr->addr[4], lladdr->addr[5],
+		get_dest_lladdr(pkt));
+
+	return NET_OK;
+}
+
+static void can_to_eth_lladdr(struct net_pkt *pkt, struct net_if *eth_iface,
+			      bool bcast)
+{
+	u16_t src_can_addr = get_src_lladdr(pkt);
+	struct net_linkaddr *lladdr;
+
+	lladdr = net_pkt_lladdr_dst(pkt);
+	lladdr->len = sizeof(struct net_eth_addr);
+	lladdr->type = NET_LINK_ETHERNET;
+	if (bcast) {
+		lladdr->addr = (u8_t *)net_eth_broadcast_addr()->addr;
+	} else {
+		/*dest address is inline*/
+		lladdr->addr = net_pkt_cursor_get_pos(pkt);
+		net_buf_pull(pkt->buffer, sizeof(struct net_eth_addr));
+	}
+
+	lladdr = net_pkt_lladdr_src(pkt);
+	lladdr->addr = net_pkt_lladdr_src(pkt)->addr -
+		       (sizeof(struct net_eth_addr) - lladdr->len);
+	memcpy(lladdr->addr, net_if_get_link_addr(eth_iface)->addr,
+	       sizeof(struct net_eth_addr));
+	lladdr->addr[4] = (lladdr->addr[4] & 0xC0) | (src_can_addr >> 8U);
+	lladdr->addr[5] = src_can_addr & 0xFF;
+	lladdr->len = sizeof(struct net_eth_addr);
+	lladdr->type = NET_LINK_ETHERNET;
+}
+
+static void forward_can_frame(struct net_pkt *pkt, struct net_if *eth_iface)
+{
+	net_pkt_set_iface(pkt, eth_iface);
+	net_if_queue_tx(eth_iface, pkt);
+}
+
+enum net_verdict translate_to_eth_frame(struct net_pkt *pkt)
+{
+	struct net_linkaddr *dest_addr = net_pkt_lladdr_dst(pkt);
+	struct net_linkaddr *src_addr = net_pkt_lladdr_src(pkt);
+	struct net_pkt *clone_pkt;
+	struct net_if *eth_iface;
+
+	/*This frame is for the CANBUS interface itself*/
+	if (!dest_is_translator(pkt) && !dest_is_mcast(pkt)) {
+		return NET_CONTINUE;
+	}
+
+	eth_iface = net_if_get_first_by_type(&NET_L2_GET_NAME(ETHERNET));
```

Comment:
Initially ok, but does not work if there are multiple ethernet interfaces (like in VLAN case).

Reply:
True. Maybe use a Kconfig for that?

Reply:
Lets do it like this in version 1, and change it to support multiple interfaces in version 2. Otherwise there is just too many changes going around.